### PR TITLE
Expose closeModalBox to Coveo.options JSUI-2246

### PR DESCRIPTION
**Pull request by jeremierobert-coveo**
Block SFINT-1844

Was added so we can set this option with a simple
```javascript
Coveo.option(searchInterfaceNode, {
  SearchInterface: {
    closeModalBox: false
  }
})
```

[Deploy](https://dashboard.heroku.com/pipelines/a3535101-5bbf-4a5b-a909-47fcf8c9f149)

### CODE REVIEW

File: src/ui/SearchInterface/SearchInterface.ts [hunk 1]
```diff
@@ -69,6 +69,7 @@ export interface ISearchInterfaceOptions {
   allowQueriesWithoutKeywords?: boolean;
   responsiveMediumBreakpoint?: number;
   responsiveSmallBreakpoint?: number;
+  closeModalBox?: boolean;
```
Comment (olamothe) [inline src/ui/SearchInterface/SearchInterface.ts]: I see the problem, but I would not pass it as an option of the search interface. This is not something that "normal" people will want to interact with, and adding an option globally like that should always be a "last resort" kind of thing.

Could we instead have a method on the query controller that you could use to switch a flag or something ?

It should also be properly tested, which it is not, currently.

Comment (jeremierobert-coveo) [reply src/ui/SearchInterface/SearchInterface.ts]: Would you mind if i inject ModalBox instead of just using it as an import ? (Was blocking my UTs)

Comment (jeremierobert-coveo) [reply src/ui/SearchInterface/SearchInterface.ts]: Also, a public method is not a bit dirty compared to an option ?
The option is also already public and can be passed to executeQuery.
Thus, put it as a component option can make sense ?

File: unitTests/controllers/QueryControllerTest.ts [hunk 2]
```diff
@@ -101,6 +101,54 @@ export function QueryControllerTest() {
...
+    describe('closeModalBox', () => {
+      let modalBoxCloseSpy: jasmine.Spy;
+      beforeEach(() => {
+        modalBoxCloseSpy = jasmine.createSpy('modal');
+        test.cmp.modalBox = {
+          close: modalBoxCloseSpy
+        };
+      });
+
+      afterEach(() => {
```
Comment (olamothe) [inline unitTests/controllers/QueryControllerTest.ts]: I don't think those are necessary.

File: unitTests/controllers/QueryControllerTest.ts [hunk 3]
```diff
@@ -101,6 +101,54 @@ export function QueryControllerTest() {
...
+        expect(modalBoxCloseSpy).toHaveBeenCalled();
+      });
+
+      it('should not close the modals if the closeModalBox is set to false', () => {
+        test.cmp.closeModalBox = false;
+        test.cmp.executeQuery();
+        expect(modalBoxCloseSpy).not.toHaveBeenCalled();
+      });
+
+      it('should close the modals if the closeModalBox is unset', () => {
```
Comment (olamothe) [inline unitTests/controllers/QueryControllerTest.ts]: If 2 different values are set on both the component (eg: false), and on the executeQuery options (eg: true), which one should have precedence ?

I'm guessing it's important in your case, and as such there should be an explicit test that check for it.

File: src/controllers/QueryController.ts [hunk 4]
```diff
@@ -168,10 +169,11 @@ export class QueryController extends RootComponent {
    * @returns {Promise<IQueryResults>}
    */
   public executeQuery(options?: IQueryOptions): Promise<IQueryResults> {
+    this.closeModalBox = options && options.closeModalBox != null ? options.closeModalBox : this.closeModalBox;
```
Comment (olamothe) [inline src/controllers/QueryController.ts]: I'd move this in a dedicated method, so that the code becomes something like : 

```
this.closeModalBoxIfNeeded(queryOptions)
```

or 

```
if (this.shouldCloseModalBox()) {
 this.modalBox.close()
}
```

Easier to read.

File: unitTests/controllers/QueryControllerTest.ts (hunk 3 above)
Comment (samisayegh) [inline unitTests/controllers/QueryControllerTest.ts]: I think a better test description would be `by default` instead of `if the closeModalBox is unset`. So it would become:

`it('should close the modals by default')`

File: src/controllers/QueryController.ts [hunk 5]
```diff
@@ -79,7 +79,6 @@ interface ILastQueryLocalStorage {
 class DefaultQueryOptions implements IQueryOptions {
   searchAsYouType = false;
   beforeExecuteQuery: () => void;
-  closeModalBox = true;
```
Comment (samisayegh) [inline src/controllers/QueryController.ts]: Why did you remove this line?

File: src/controllers/QueryController.ts [hunk 6]
```diff
@@ -95,6 +94,8 @@ export class QueryController extends RootComponent {
   static ID = 'QueryController';
   public historyStore: CoveoAnalytics.HistoryStore;
   public firstQuery: boolean;
+  public modalBox = ModalBox;
```
Comment (ThibodeauJF) [inline src/controllers/QueryController.ts]: Do you expose the `modalBox` for easier testing?
